### DISCUSSION TITLE
FilterOnObservable how use?

### DISCUSSION THREAD

**Discussion by POMXARK** [reactivemarbles/DynamicData#661] [Q&A]
Hello, the operator is implemented in the new version (DynamicData 7.9.5) FilterOnObservable

But the example uses self-written

DynamicData.Samplz.Examples

        public FilterObservableViewModel()
        {
            //Load available players
            CreateFootballerList().Connect()
                .FilterOnObservable(person => person.IncludedChanged, included => !included)
                .ObserveOnDispatcher()
                .Bind(out _availablePlayers)
                .Subscribe();
        }

Does the new method give the same result, how can I bring an example to a new look. Tell me, thanks.

I use [AvaloniaUI/Avalonia] and Entity Framework

        public static IObservable<IChangeSet<TObject>> FilterOnObservable<TObject, TValue>(this IObservable<IChangeSet<TObject>> source,
            Func<TObject, IObservable<TValue>> observableSelector,
            Func<TValue, bool> predicate
            )

convert to 

public static IObservable<IChangeSet<TObject>> FilterOnObservable<TObject>(this IObservable<IChangeSet<TObject>> source, Func<TObject, IObservable<bool>> objectFilterObservable, TimeSpan? propertyChangedThrottle = null, IScheduler? scheduler = null)

**Comment by SuperJMN** [comment]
Please, take a look to the following example:

```csharp
void Main()
{
	var m1 = new Model(1);
	
	var models = new[] 
	{
		m1,
		new Model(2), 
		new Model(3)
	};
	
	var changeSet = models.AsObservableChangeSet();
	
	changeSet
		.FilterOnObservable(x => x.IsActive)
		.Bind(out var filteredCollection)
		.Subscribe();
		
	m1.Activate();
	
	// Prints 1	
	filteredCollection.ToList().ForEach(x => Console.WriteLine(x.Id));	
}

class Model 
{
	private BehaviorSubject<bool> isActive = new(false);
	
	public Model(int id)
	{
		Id = id;
	}
	
	public int Id { get; }

	public IObservable<bool> IsActive => isActive.AsObservable();

	public override string ToString()
	{
		return Id.ToString();
	}

	internal void Activate()
	{
		isActive.OnNext(true);
	}
}
```

I think that you will understand how it works quickly. 
Basically, it filters items based on the last bool emitted by the observable you specify in the lambda.

**Comment by dwcullop** [comment]
> ```
>     public static IObservable<IChangeSet<TObject>> FilterOnObservable<TObject, TValue>(this IObservable<IChangeSet<TObject>> source,
>         Func<TObject, IObservable<TValue>> observableSelector,
>         Func<TValue, bool> predicate
>         )
> ```

As @SuperJMN showed in his example, the library requires a factory that returns `IObservable<bool>` that indicates if that value should pass the filter.   However, if you want one with this prototype, you can adapt the one in the library to the one you want as follows:

```csharp
public static IObservable<IChangeSet<TObject>> FilterOnObservable<TObject, TValue>(this IObservable<IChangeSet<TObject>> source, 
    Func<TObject, IObservable<TValue>> observableSelector, 
    Func<TValue, bool> predicate) =>
    source.FilterOnObservable(obj => observableSelector(obj).Select(val => predicate(val)));
```

**Comment by Snailya** [comment]
Where to find DynamicData.Samplz.Examples?

**Comment by JakenVeina** [comment]
[This](https://github.com/RolandPheasant/DynamicData.Snippets) is probably what you're looking for.
